src/kernel/scheduler.zig

```diff
@@ -0,0 +1,583 @@
+    /// state as described in arch.CpuState struct.
+    ///
+    /// Arguments:
+    ///     IN comptime entry_point: var - The entry point into the task. This must be a function.
+    ///
+    /// Return: *Task
+    ///     Pointer to an allocated task. This will then need to be added to the task queue.
+    ///
+    /// Error: std.mem.Allocator.Error
+    ///     OurOfMemory - If there is no more memory to allocate. Any memory or PID allocated will
+    ///                   be freed on return.
+    ///
+    pub fn init(comptime entry_point: var) Allocator.Error!*Task {
+        // Ensure the entry point is a function
+        comptime assert(@typeInfo(@TypeOf(entry_point)) == .Fn);
+
+        var task = try alloc.create(Task);
+        errdefer alloc.destroy(task);
+
+        task.pid = allocatePid();
+        errdefer freePid(task.pid);
+
+        // Will need to copy the cpu state onto the stack from eip onwards
+        task.stack = try alloc.alloc(u32, STACK_SIZE);
+
+        // TODO Will need to add the exit point
+        // Set up everything as a kernel task
+        task.stack[STACK_SIZE - 18] = gdt.KERNEL_DATA_OFFSET; // ss
+        task.stack[STACK_SIZE - 17] = gdt.KERNEL_DATA_OFFSET; // gs
+        task.stack[STACK_SIZE - 16] = gdt.KERNEL_DATA_OFFSET; // fs
+        task.stack[STACK_SIZE - 15] = gdt.KERNEL_DATA_OFFSET; // es
+        task.stack[STACK_SIZE - 14] = gdt.KERNEL_DATA_OFFSET; // ds
+
+        task.stack[STACK_SIZE - 13] = 0; // edi
+        task.stack[STACK_SIZE - 12] = 0; // esi
+        // End of the stack
+        task.stack[STACK_SIZE - 11] = @ptrToInt(&task.stack[STACK_SIZE - 1]); // ebp
+        task.stack[STACK_SIZE - 10] = 0; // esp (temp) this won't be popped by popa bc intel is dump XD
+
+        task.stack[STACK_SIZE - 9] = 0; // ebx
+        task.stack[STACK_SIZE - 8] = 0; // edx
+        task.stack[STACK_SIZE - 7] = 0; // ecx
+        task.stack[STACK_SIZE - 6] = 0; // eax
+
+        task.stack[STACK_SIZE - 5] = 0; // int_num
+        task.stack[STACK_SIZE - 4] = 0; // error_code
+
+        task.stack[STACK_SIZE - 3] = @ptrToInt(entry_point); // eip
+        task.stack[STACK_SIZE - 2] = gdt.KERNEL_CODE_OFFSET; // cs
+        task.stack[STACK_SIZE - 1] = 0x202; // eflags
+
+        // Point to top of the created stack above
+        task.esp = @ptrToInt(&task.stack[STACK_SIZE - 18]);
+
+        return task;
+    }
+
+    ///
+    /// Destroy the task. This will release the allocated PID and free the stack and self.
+    ///
+    /// Arguments:
+    ///     IN/OUT self: *Self - The pointer to self.
+    ///
+    pub fn destroy(self: *Self) void {
+        freePid(self.pid);
+        alloc.free(self.stack);
+        alloc.destroy(self);
+    }
+
+    ///
+    /// Save a new stack pointer for this task.
+    ///
+    /// Arguments:
+    ///     IN/OUT self: *Self - The pointer to self.
+    ///     IN esp: u32        - The new stack pointer to be saved.
+    ///
+    pub fn saveEsp(self: *Self, esp: u32) void {
```

Comment:
I think this should also be moved to the arch file.

Reply:
Task struct is being moved to arch file